### PR TITLE
New UI: tips

### PR DESCRIPTION
**Pull request by breville**
This implements a generic React component called `UITips` which provides some pop-up tips on any page.

It's built to support the first two use cases implemented in this PR: /home and /courses for teachers.  

On /home, the teacher will see a "before" dialog, then a couple "initial" tips which can be dismissed, followed by an "after" dialog.  Once both "initial" tips have been dismissed, they won't be seen again due to a property written to the user object.  However, clicking on the Code.org logo (which happens to not be implemented in React) will always "trigger" a third tip. 

On /courses, the teacher will see a single "initial" tip which can be dismissed.  Once it has been dismissed, it won't be seen again.

The whole tips UI is hidden when the browser width is <= 970px.

Context: this change was written right before the new UI release and is unfortunately not tested.  It's also got an architectural weakness in that the state manager `UITips` wraps the individuals tips, whereas it would be better if each tip existed independently throughout the DOM so that they could be located next to the element they are pointing at.  In the current implementation, it's difficult for the manager to place the tips beside specific elements in the DOM, using absolute coordinates, because their locations aren't known until the whole tree is rendered.  A better implementation would have had a global state (e.g. in Redux) managed by a tips manager, while individual tips would be situated at the right places in the DOM, and would render depending on that global state.  As it is, the way the tip is shown on /courses is simply by locating `UITips` immediately above the course explorer.

Aside: this change also prevents overlapping dialogs from showing for teachers on /home.  Where previously both the terms and school info dialogs could be shown at once, now only terms is shown if necessary, otherwise school info, otherwise tips.

### CODE REVIEW

**Comment by breville** [comment]
cc @tanyaparker

File: dashboard/app/controllers/api/v1/users_controller.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ class Api::V1::UsersController < ApplicationController
 
   def load_user
     user_id = params[:user_id]
-    if user_id != 'me' && user_id != current_user.id
```
Comment (breville) [inline dashboard/app/controllers/api/v1/users_controller.rb:9]: @pcardune Not sure whether this was working before, but in my testing `user_id` was a `String` while `current_user.id` was a `Fixnum` and the comparison was not succeeding.

File: apps/src/templates/teacherHomepage/UiTip.jsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    this.props.closeClicked(index);
+  },
+
+  render() {
+    const { index, position, text } = this.props;
+
+    return (
+      <div>
+        <div
+          className="arrow_box_up"
```
Comment (Bjvanminnen) [inline apps/src/templates/teacherHomepage/UiTip.jsx]: Is there a reason we use a className instead of inline styles?

Comment (breville) [reply apps/src/templates/teacherHomepage/UiTip.jsx]: Yes, for the `:after` selector.  Wasn't sure how to do it in React.  

https://github.com/code-dot-org/code-dot-org/pull/15463/files#diff-2f5fadbda247dcceb98bc8410d0ecaa3R2880

File: apps/src/templates/teacherHomepage/UiTips.jsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            onConfirm={this.afterDialogConfirm}
+          />
+        )}
+
+      </div>
+    );
+  }
+});
+
+export default UiTips;
```
Comment (Bjvanminnen) [inline apps/src/templates/teacherHomepage/UiTips.jsx]: This class has quite a bit of logic in it, and no testing that I see.

File: apps/src/templates/teacherHomepage/UiTips.jsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    let newState = {...this.state, showingDialog: null};
+    this.setState(newState);
+  },
+
+  afterDialogConfirm() {
+    if (this.props.afterDialog.onConfirm.action === "url") {
+      window.open(this.props.afterDialog.onConfirm.url, "_blank");
+    }
+    let newState = {...this.state, showingDialog: null};
+    this.setState(newState);
```
Comment (Bjvanminnen) [inline apps/src/templates/teacherHomepage/UiTips.jsx]: here (and elsewhere), you don't need to set the full state every time. If you replace the previous two lines with 
```jsx
this.setState({showingDialog: null});
```
You end up with the same result (React does a merge of current state and whatever you're passing in to setState)

File: apps/src/templates/teacherHomepage/UiTips.jsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+import Dialog from '../Dialog';
+
+const UiTips = React.createClass({
+  propTypes: {
+    userId: React.PropTypes.number,
+    tipId: React.PropTypes.string,
+    showInitialTips: React.PropTypes.bool,
+    beforeDialog: React.PropTypes.object,
+    afterDialog: React.PropTypes.object,
+    tips: React.PropTypes.array
```
Comment (Bjvanminnen) [inline apps/src/templates/teacherHomepage/UiTips.jsx]: might be good to specify the shape of a `tip`

File: apps/src/templates/teacherHomepage/Courses.jsx [hunk 6]
```diff
@@ -51,10 +54,22 @@ const Courses = React.createClass({
   },
 
   render() {
-    const { courses, isEnglish, isTeacher, codeOrgUrlPrefix, isSignedOut } = this.props;
+    const { courses, isEnglish, isTeacher, codeOrgUrlPrefix, isSignedOut, userId, showInitialTips } = this.props;
 
     return (
       <div>
+        {isEnglish && isTeacher && (
```
Comment (Bjvanminnen) [inline apps/src/templates/teacherHomepage/Courses.jsx]: Non-english teachers dont get the benefit of tips?

File: apps/src/templates/teacherHomepage/UiTip.jsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+          onClick={this.closeClicked.bind(this, index)}
```
Comment (Bjvanminnen) [inline apps/src/templates/teacherHomepage/UiTip.jsx]: shouldnt need to bind closeClicked (since this is using `createClass` which autobinds class methods)

File: apps/src/templates/teacherHomepage/UiTip.jsx (hunk 2 above)
Comment (Bjvanminnen) [reply apps/src/templates/teacherHomepage/UiTip.jsx]: Do we need to use an after selector vs. just sticking a styled DOM element after this one? If this answer is that we do need to use :after, I would advocated for still having the non-after stuff inlined (but dont feel super strongly).

File: dashboard/app/controllers/api/v1/users_controller.rb (hunk 1 above)
Comment (pcardune) [reply dashboard/app/controllers/api/v1/users_controller.rb:9]: The original use case only made use of `'me'` so the other part may have gone untested.

File: dashboard/app/controllers/api/v1/users_controller.rb [hunk 8]
```diff
@@ -35,4 +35,13 @@ def post_using_text_mode
 
     render json: {using_text_mode: !!@user.using_text_mode}
   end
+
+  # POST /api/v1/users/<user_id>/post_ui_tip_dismissed
+  def post_ui_tip_dismissed
```
Comment (pcardune) [inline dashboard/app/controllers/api/v1/users_controller.rb:40]: There are already tests for `get/post_using_text_mode`, so you should add some additional tests for `post_ui_tip_dismissed`

File: apps/src/templates/studioHomepages/UiTips.jsx [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+          )
+        )))}
+
+        {this.props.beforeDialog && (
+          <Dialog
+            isOpen={this.state.showingDialog === "before"}
+            title={this.props.beforeDialog.title}
+            body={this.props.beforeDialog.body}
+            confirmText={this.props.beforeDialog.confirm}
+            cancelText={this.props.beforeDialog.cancel}
```
Comment (pcardune) [inline apps/src/templates/studioHomepages/UiTips.jsx:176]: You can use a spread operator in react components to simplify this a bit:

```jsx
<Dialog
  isOpen={this.state.showingDialog === "before"}
  {...this.props.beforeDialog}
  onCancel={this.beforeDialogCancel}
  onConfirm={this.beforeDialogConfirm}
/>
```

File: apps/src/templates/studioHomepages/UiTips.jsx [hunk 10]
```diff
@@ -0,0 +1,176 @@
+import React from 'react';
+import UiTip from './UiTip';
+import Dialog from '../Dialog';
+
+const UiTips = React.createClass({
+  propTypes: {
+    userId: React.PropTypes.number,
+    tipId: React.PropTypes.string,
+    showInitialTips: React.PropTypes.bool,
+    beforeDialog: React.PropTypes.object,
```
Comment (pcardune) [inline apps/src/templates/studioHomepages/UiTips.jsx:19]: You can specify these more than just being an object. In fact, it's easy to just reference the propTypes of the Dialog component:

```jsx
  propTypes: {
    ...
    beforeDialog: React.PropTypes.shape(Dialog.propTypes),
    afterDialog: React.PropTypes.shape(Dialog.propTypes),
  }
```

File: apps/src/sites/studio/pages/home/_homepage.js [hunk 11]
```diff
@@ -3,16 +3,48 @@ import React from 'react';
...
+  const showUiTips = homepageData.showuitips;
+  const userId = homepageData.userid;
+  const showInitialTips = !homepageData.initialtipsdismissed;
 
   ReactDOM.render (
     <div>
+      {showUiTips && (
+        <UiTips
+          userId={userId}
+          tipId={"homepage_header"}
```
Comment (pcardune) [inline apps/src/sites/studio/pages/home/_homepage.js]: string attributes do not need to be wrapped in curly braces.

File: apps/src/templates/studioHomepages/UiTip.jsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  },
+
+  render() {
+    const { index, position, text } = this.props;
+
+    return (
+      <div>
+        <div
+          className="arrow_box_up"
+          onClick={this.closeClicked.bind(this, index)}
```
Comment (pcardune) [inline apps/src/templates/studioHomepages/UiTip.jsx]: when using `React.createClass`, all the methods are automatically bound to `this` for you. So you shouldn't need to bind it here.

File: apps/src/templates/studioHomepages/UiTip.jsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+const UiTip = React.createClass({
+  propTypes: {
+    index: React.PropTypes.number,
+    position: React.PropTypes.object,
+    text: React.PropTypes.string,
+    closeClicked: React.PropTypes.func.isRequired
+  },
+
+  closeClicked(index) {
+    this.props.closeClicked(index);
```
Comment (pcardune) [inline apps/src/templates/studioHomepages/UiTip.jsx:13]: as far as I can tell, the only possible value of `index` is `this.props.index`. So you are better off getting rid of the `index` parameter and just referencing `this.props.index` directly.

File: apps/src/templates/studioHomepages/UiTips.jsx [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+    // Once all the initial tips are dismissed, tell the server to update the user
+    // so it's recorded.
+    if (this.state.showInitialTips) {
+      let tipRemaining = false;
+      newState.tipsShowing.forEach(tipShowing => {
+        if (tipShowing) {
+          tipRemaining = true;
+        }
+      });
```
Comment (pcardune) [inline apps/src/templates/studioHomepages/UiTips.jsx:63]: this whole bit can be simplified with [`Array.some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some?v=control):

```jsx
let tipRemaining = newState.tipsShowing.some(f => f);
```

File: apps/src/templates/studioHomepages/UiTips.jsx [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+        newState = {...newState, showInitialTips: false};
+      }
+    }
+
+    this.setState(newState);
+  },
+
+  afterDialogCancel() {
+    let newState = {...this.state, showingDialog: null};
+    this.setState(newState);
```
Comment (pcardune) [inline apps/src/templates/studioHomepages/UiTips.jsx:86]: You don't need to copy the entire state when calling `setState`. `setState` will do the merge for you. So this can just be:

```jsx
this.setState({showingDialog: null});
```

File: apps/src/templates/studioHomepages/UiTips.jsx [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+      if (tip.type === "triggered") {
+        $(`#${tip.triggerId}`).click(e => {
+          if (!this.state.showInitialTips) {
+            let newTipsShowing = [...this.state.tipsShowing];
+            newTipsShowing[index] = true;
+            const newState = {...this.state, tipsShowing: newTipsShowing};
+            this.setState(newState);
+          }
+          e.preventDefault();
+        });
```
Comment (pcardune) [inline apps/src/templates/studioHomepages/UiTips.jsx:137]: Note that if this component ever gets re-rendered or unrendered, you will potentially get some weird behavior... for example:

1. When component is unmounted, the event handlers will be left around and they will throw errors when clicked.
2. When the component is re-rendered, and if `this.props.tips` has changed, any new tips won't have event handlers attached, and any old tips won't have their event handlers removed.

Or in short: if you are going to break out of react to do things directly with the DOM, make sure you fully implement the react lifecycle methods to keep the DOM in sync _or_ leave a really big comment explaining that this component has side effects and will not behave like normal react components. Specifically the other lifecycle methods you would need to implement here are `componentWillUnmount` and `componentDidUpdate`.

File: apps/src/templates/studioHomepages/UiTip.jsx [hunk 17]
```diff
@@ -13,12 +14,11 @@ const UiTip = React.createClass({
...
 
   render() {
-    const { index, position, text } = this.props;
-
+    const { index, position, text, arrowDirection } = this.props;
     return (
       <div>
         <div
-          className="arrow_box_up"
+          className={`arrow_box_${arrowDirection}`}
```
Comment (Bjvanminnen) [inline apps/src/templates/studioHomepages/UiTip.jsx]: I really dislike dynamically generated classnames, as they become impossible to search for. I'd much rather see something like
```jsx
className={(arrowDirection === 'up') ? 'arrow_box_up' : 'arrow_box_down'}
```

File: apps/src/templates/studioHomepages/UiTips.jsx [hunk 18]
```diff
@@ -32,7 +33,8 @@ const UiTips = React.createClass({
     return {
       showInitialTips: showInitialTips,
       tipsShowing: tipsShowing,
-      showingDialog: showingDialog
+      showingDialog: showingDialog,
+      mobileWidth: $(window).width() <= 970
```
Comment (Bjvanminnen) [inline apps/src/templates/studioHomepages/UiTips.jsx]: Is this a constant that should be defined in some shared location?

File: dashboard/app/views/courses/index.html.haml [hunk 19]
```diff
@@ -13,7 +13,7 @@
   courses_data[:studentscount] = students_count
   courses_data[:codeorgurlprefix] = CDO.code_org_url
   courses_data[:signedout] = @is_signed_out
-  courses_data[:initialtipsdismissed] = current_user.try(:ui_tip_dismissed_teacher_courses)
+  courses_data[:initialtipsdismissed] = current_user ? current_user.ui_tip_dismissed_teacher_courses : nil
```
Comment (Bjvanminnen) [inline dashboard/app/views/courses/index.html.haml]: why this change?

File: apps/src/templates/teacherHomepage/Courses.jsx (hunk 6 above)
Comment (breville) [reply apps/src/templates/teacherHomepage/Courses.jsx]: Not for this control which is also hidden for non-en teachers.

File: dashboard/app/controllers/api/v1/users_controller.rb (hunk 8 above)
Comment (breville) [reply dashboard/app/controllers/api/v1/users_controller.rb:40]: Added one.

File: apps/src/templates/studioHomepages/UiTips.jsx (hunk 16 above)
Comment (breville) [reply apps/src/templates/studioHomepages/UiTips.jsx:137]: Under current design, this lives for the duration of the page, but I will add a comment.

File: dashboard/app/models/user.rb [hunk 20]
```diff
@@ -109,6 +109,8 @@ class User < ActiveRecord::Base
     races
     using_text_mode
     last_seen_school_info_interstitial
+    ui_tip_dismissed_homepage_header
+    ui_tip_dismissed_teacher_courses
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb:113]: Do we have a plan to sunset these serialized properties? Presumably, at some point we want to stop showing those tooltips and remove this serialized attribute? Assuming this is the case, do we know a rough timeline as to when this should happen?

Comment (breville) [reply dashboard/app/models/user.rb:113]: Good question.  @tanyaparker should make the call on that.

Comment (tanyaparker) [reply dashboard/app/models/user.rb:113]: I want to keep them up through Hour of Code (remove January) or when 90% of teachers have seen it. So we can take them down after whichever comes first.
